qiskit/qobj/_qobj.py

```diff
@@ -0,0 +1,209 @@
+        init_args = tuple(getattr(self, key) for key in self.REQUIRED_ARGS)
+        extra_args = {key: value for key, value in self.__dict__.items()
+                      if key not in self.REQUIRED_ARGS}
+        return self.__class__, init_args, extra_args
+
+
+class QObj(QObjItem):
+    """Representation of a QObj.
+
+    Attributes:
+        id (str): QObj identifier.
+        config (QObjConfig): config settings for the QObj.
+        circuits (list[QObjExperiment]): list of experiments.
+        type (str): experiment type (QASM/PULSE).
+        _version (str): QObj version.
+    """
+
+    REQUIRED_ARGS = ['id', 'config', 'circuits']
+
+    def __init__(self, id, config, circuits, **kwargs):
+        # pylint: disable=redefined-builtin,invalid-name
+        self.id = id
+        self.config = config
+        self.circuits = circuits
+
+        self.type = QObjType.QASM.value
+        self._version = QOBJ_VERSION
+
+        super().__init__(**kwargs)
+
+
+class QObjConfig(QObjItem):
+    """Configuration for a QObj.
+
+    Attributes:
+        max_credits (int): number of credits.
+        shots (int): number of shots.
+        backend_name (str): name of the backend.
+    """
+    REQUIRED_ARGS = ['max_credits', 'shots', 'backend_name']
+
+    def __init__(self, max_credits, shots, backend_name, **kwargs):
+        self.max_credits = max_credits
+        self.shots = shots
+        self.backend_name = backend_name
+
+        super().__init__(**kwargs)
+
+
+class QObjExperiment(QObjItem):
+    """Quantum experiment represented inside a QObj.
+
+    Attributes:
+        name (str): name of the experiment.
+        config (QObjExperimentConfig): config settings for the experiment.
+        compiled_circuit (QObjCompiledCircuit): list of instructions
+        compiled_circuit_qasm (str)
+    """
+    REQUIRED_ARGS = ['name', 'config', 'compiled_circuit',
+                     'compiled_circuit_qasm']
+
+    def __init__(self, name, config, compiled_circuit, compiled_circuit_qasm,
+                 **kwargs):
+        self.name = name
+        self.config = config
+        self.compiled_circuit = compiled_circuit
+        self.compiled_circuit_qasm = compiled_circuit_qasm
+
+        super().__init__(**kwargs)
+
+
+class QObjInstruction(QObjItem):
+    """Quantum Instruction.
+
+    Attributes:
+        name(str): name of the gate.
+        qubits(list): list of qubits to apply to the gate.
+    """
+    REQUIRED_ARGS = ['name', 'qubits']
+
+    def __init__(self, name, qubits, **kwargs):
+        self.name = name
+        self.qubits = qubits
+
+        super().__init__(**kwargs)
+
+
+# TODO: Remove when new schema is in place?
+class QObjExperimentConfig(QObjItem):
+    """Configuration for a experiment.
+
+    Attributes:
+        seed (int): seed.
+        basis_gates (str): basis gates
+        coupling_map (list): coupling map
+        layout (list): layout
+    """
+    REQUIRED_ARGS = ['seed', 'basis_gates', 'coupling_map', 'layout']
```

Comment:
(Flagging this here, but it is fine if you want to fix in a future PR.)
`basis_gates` and `coupling_map` should not be part of a `QobjExperimentConfig`. These are properties of the backend, so the backend already knows about them and we shouldn't include them in the payload being sent to the backend. I have to think about `layout` more, but I think `layout` can also be deleted here.